contracts/astroport/oracle/src/contract.rs

```diff
@@ -152,38 +133,44 @@ pub fn set_asset_infos(
 pub fn update(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
     let config = CONFIG.load(deps.storage)?;
     let pair = config.pair.ok_or(ContractError::AssetInfosNotSet {})?;
-    let price_last = PRICE_LAST.load(deps.storage)?;
-
     let prices = query_cumulative_prices(deps.querier, pair.contract_addr)?;
-    let time_elapsed = env.block.time.seconds() - price_last.block_timestamp_last;
-
-    // Ensure that at least one full period has passed since the last update
-    if time_elapsed < config.period {
-        return Err(ContractError::WrongPeriod {});
-    }
 
     let mut average_prices = vec![];
-    for (asset1_last, asset2_last, price_last) in price_last.cumulative_prices.iter() {
-        for (asset1, asset2, price) in prices.cumulative_prices.iter() {
-            if asset1.equal(asset1_last) && asset2.equal(asset2_last) {
-                average_prices.push((
-                    asset1.clone(),
-                    asset2.clone(),
-                    Decimal256::from_ratio(
-                        Uint256::from(price.wrapping_sub(*price_last)),
-                        time_elapsed,
-                    ),
-                ));
+    if let Some(price_last) = PRICE_LAST.may_load(deps.storage)? {
+        let time_elapsed = env.block.time.seconds() - price_last.block_timestamp_last;
+        // Ensure that at least one full period has passed since the last update
+        if time_elapsed < config.period {
+            return Err(ContractError::WrongPeriod {});
+        }
+
+        for (asset1_last, asset2_last, price_last) in price_last.cumulative_prices.iter() {
+            for (asset1, asset2, price) in prices.cumulative_prices.iter() {
```

Comment:
wdyt about using `filter` instead of loop?

Reply:
I really wouldn't waste time on it atm. this code is taken as is from the origin — astroport dudes